Add section about destination header provider (#92)
SAP/cloud-sdk: docs/java/features/connectivity/destinations.mdx

Use `prependDestinationLoader()` to add it at the beginning if you would like it to take precedence.
Lastly use `setLoader()` to replace all existing loaders.

## Provide Headers For Destinations

### How It Works

> You can provide headers which are issued for any request over destinations.
> You can implement custom logic which defines which headers are sent under which circumstances.

Reviewer: ```suggestion
You can add headers that are sent for any request over destinations.
You can implement custom logic that defines which headers are sent under which circumstances.
```

Reviewer: I used a bit more frequent vocab for headers.

> :::note Example Usage

Reviewer: ```suggestion
:::note Example Use-case
```

> For example, you want to issue an API token via the request header to any calls made over any destination in your whole application.

Reviewer: ```suggestion
Let's imagine, you want to send an API token via a request header with any calls made over any destination in your whole application.
```

:::

> The SAP Cloud SDK offers the interface `DestinationHeaderProvider` which defines the method `List<Header> getHeaders(destination)`.

Reviewer 2: We use `DestinationProperties` as synonym for _destination_ because it's the most generic interface that would need to be implemented.

At runtime, while preparing an outbound request the SAP Cloud SDK looks up all implementations of that interface on the classpath.
For each found implementation it invokes the method `getHeaders(destination)` and appends the returned headers to the currently prepared request.

### Implement Own Header Provider

#### Implement the Interface
To implement your custom destination header provider you need to implement the interface `DestinationHeaderProvider`.

> The method `getHeaders` lets you access the destination over which the currently prepared request will be issued.

Reviewer 3: ```suggestion
The method `getHeaders` lets you access the destination properties over which the currently prepared request will be issued.
```

You can read destination properties (for example, the name and the URI) if your logic requires to determine the headers.

Here is one simple example implementation which returns the same header for each request:

```java
@Nonnull
@Override
public List<Header> getHeaders( @Nonnull final DestinationProperties destination )
{
  final Header header = new Header("API-Token", obtainApiToken());

  return Collections.singletonList(header);
}
```

#### Make Interface Implementation Accessible
The SAP Cloud SDK uses the [Service Loader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html) mechanism of Java to lookup the `DestinationHeaderProvider` implementations at runtime.

> To let your custom implementation be found, you need to provide a configuration file as resource in the folder `META-INF/services`.

Reviewer 3: Should we mention here the relative location where the META-INF folder needs to reside? For example:

  In a regular maven project, the META-INF folder would be located in `src/main/resources/`.

Author: Good idea, I added a mention about this

> To let your custom implementation be found, you need to provide a configuration file as resource in the folder `META-INF/services`.

Reviewer 2: Note: I think this is only necessary when your "application" is not deployed but packaged as dependency in another app.

Reviewer 3: Hopefully this describes my setup accurately: My implementation was located in the "application" module of my application, and this was not enough to get it working. I named the resource folder wrongly (META_INF). Only after I corrected that and named it "META-INF", was the implementation found by the ServiceLoader. I am not sure if what you say is still true, since the "application" module is like a dependency.

Author: Such a config file is always needed. 
Check this:
- SO: https://stackoverflow.com/a/10304246/4790468
- Official guide: https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html (does not mention that there is a difference between the implementation being in the same jar or in another jar)

> Name this resource file as the interface `com.sap.cloud.sdk.cloudplatform.connectivity.DestinationHeaderProvider` and mention the fully-qualified class name of your custom implementation.

Reviewer 3: ```suggestion
Name this resource file as the interface `com.sap.cloud.sdk.cloudplatform.connectivity.DestinationHeaderProvider` and mention the fully-qualified class name of your custom implementation.
You can refer to several implementations by putting their fully qualified class names on separate lines.
```

Use `prependDestinationLoader()` to add it at the beginning if you would like it to take precedence.
Lastly use `setLoader()` to replace all existing loaders.

## Provide Headers For Destinations

### How It Works

You can add headers that are sent for any request over destinations.
You can implement custom logic that defines which headers are sent under which circumstances.

:::note Example Use-case
Let's imagine, you want to send an API token via a request header with any calls made over any destination in your whole application.
:::

The SAP Cloud SDK offers the interface `DestinationHeaderProvider` which defines the method `List<Header> getHeaders(destination)`.

> At runtime, while preparing an outbound request, the SAP Cloud SDK looks up all implementations of that interface on the classpath.
> For each found implementation it invokes the method `getHeaders(destination)` and appends the returned headers to the currently prepared request.

Reviewer 3: Just occurred to me: This is misleading, in that only those implementations will be used, that are referenced in the META-INF resource.

```suggestion
At runtime, while preparing an outbound request, the SAP Cloud SDK will use all implementations of that interface which have been registered via the Java Service mechanism (see below).
For each registered implementation it invokes the method `getHeaders(destination)` and appends the returned headers to the currently prepared request.
```

Author: good catch @stippi2 !
